threading/basic_threading/source/basic_threading.c

```diff
@@ -0,0 +1,218 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <time.h>
+#include <gccore.h>
+#include <wiiuse/wpad.h>
+#include <ogc/lwp.h>
+
+// Framebuffer and Render Mode.
+static void *xfb = NULL;
+static GXRModeObj *rmode = NULL;
+
+// Initialization.
+void Init();
+
+// Example Data Type
+typedef struct {
+    int id;     // ID.
+    int tg;     // Variable Target.
+    int var;    // Some Variable.
+} demo_data_t;
+
+// Thread Delay Time
+#define THREAD_SLEEP_TIME 1000*1000 // One Second
+
+// Thread Context Handles.
+static lwp_t handle_a = LWP_THREAD_NULL;
+static lwp_t handle_b = LWP_THREAD_NULL;
+
+// Example Thread
+//  void* arg - Data to pass to the thread.
+static void* demo_thread(void* arg) {
+    // Get the data
+    demo_data_t* demo_data;
+    if(arg != NULL) demo_data = (demo_data_t*)arg;
+    else return NULL;
+
+    // Pre-Loop Code.
+    printf("%s(#%d): Started!!! Reach Target: %d.\n",
+        __FUNCTION__,
+        demo_data->id,
+        demo_data->tg
+    );
+
+    // Thread Loop
+    bool thread_done = false;
+    while(!thread_done) {
+        // Thread Code
+        printf("%s(#%d): var = %d.\n",
+            __FUNCTION__,
+            demo_data->id,
+            demo_data->var
+        );
+
+        // Increment the variable until it reaches target.
+        demo_data->var++;
+        if(demo_data->var > demo_data->tg) {
+            thread_done = true; // Reached Target.
+        }
+
+        // Thread Delay
+        usleep(THREAD_SLEEP_TIME);
+    }
+
+    // Finished!
+    printf("%s(#%d): Finished!\n",
+        __FUNCTION__,
+        demo_data->id
+    );
+    return NULL; // Thread Done!
+}
+
+// Main Code.
+int main(int argc, char** argv) {
+    // Initialize.
+    Init();
+
+    // Create some demo data.
+    demo_data_t* demo_data = malloc(sizeof(demo_data_t) * 2); // Data for 2 threads.
+    for(int idx=0; idx < 2; idx++) {
+        printf("Setting up demo data[%d]...\n", idx);
+        demo_data[idx].id = idx+1;          // Set ID
+        demo_data[idx].tg = rand() % 10;    // Random Target Value.
+    }
+
+    // Init Thread Handles.
+    // Thread Handle A
+    if(LWP_CreateThread(
+        &handle_a,              // Thread Handle
+        demo_thread,            // Thread Function
+        (void*)&demo_data[0],   // Argument Data
+
+        NULL,0,                 // Stack and Stack Size.
+        LWP_PRIO_HIGHEST        // Thread Priority.
+    ) != LWP_SUCCESSFUL) fprintf(stderr, "Error setting up demo HandleA\n");
+
+    // Thread Handle B
+    if(LWP_CreateThread(
+        &handle_b,
+        demo_thread,
+        (void*)&demo_data[1],
+
+        NULL,0,
+        LWP_PRIO_HIGHEST
+    ) != LWP_SUCCESSFUL) fprintf(stderr, "Error setting up demo HandleB\n");
+
+    // Main Loop.
+    bool quit=false;
+    while(!quit) {
+        // Scan WPAD
+        WPAD_ScanPads();
+        s32 pressed = WPAD_ButtonsDown(WPAD_CHAN_0);
+        if(pressed & WPAD_BUTTON_HOME) quit=true;
```

Comment:
to complete a thread example, you also need code to stop (join) the threads. please add that :)

Reply:
I'll get right on that